[DataGrid] Support advanced server-side pagination use cases (#12474)
mui/mui-x: docs/data/data-grid/pagination/pagination.md

1. Row count is available (known)
2. Row count is not available (unknown)
3. Row count is available but is not accurate and may update later on (estimated)

:::warning
The `rowCount` prop is used in server-side pagination mode to inform the DataGrid about the total number of rows in your dataset.
This prop is ignored when the `paginationMode` is set to `client`, i.e. when the pagination is handled on the client-side.
:::

You can achieve each of these use-cases by setting the values of `rowCount`, `paginationMeta.hasNextPage`, and `estimatedRowCount` props. The following table summarizes the use of these props in each of these scenarios.

|                              | Known row count | Unknown row count | Estimated row count |
| :--------------------------- | :-------------- | :---------------- | :------------------ |
| `rowCount`                   | `number`        | `-1`              | `-1`                |
| `paginationMeta.hasNextPage` | —               | `boolean`         | `boolean`           |
| `estimatedRowCount`          | —               | —                 | `number`            |

The following examples demonstrate each of these scenarios:

> #### Known row count

Reviewer: Could we keep the pagination buttons active while loading, as it currently is on `master`?
https://mui.com/x/react-data-grid/pagination/#basic-implementation
This allows switching pages faster if necessary.

Reviewer: Thanks for linking to that issue, I wasn't aware of it.

> I am trying to use the datagrid in a way where server side pagination using cursors provides us the ability to go to the next page and reliably get the next results.

From what I can observe in https://deploy-preview-12474--material-ui-x.netlify.app/x/react-data-grid/pagination/#cursor-based-pagination, even though the pagination buttons aren't disabled while loading, I can't skip pages by clicking while loading 👍🏻

```jsx
const rowCountRef = React.useRef(pageInfo?.totalRowCount || 0);

const rowCount = React.useMemo(() => {
  if (pageInfo?.totalRowCount !== undefined) {
    rowCountRef.current = pageInfo.totalRowCount;
  }
  return rowCountRef.current;
}, [pageInfo?.totalRowCount]);

<DataGrid rowCount={rowCount} />;
```

{{"demo": "ServerPaginationGrid.js", "bg": "inline"}}

#### Unknown row count

Pass the props to the Grid as explained in the table above to handle the case when the actual row count is not initially available.

The value of the `hasNextPage` variable might become `undefined` during loading if it's handled by some external fetching hook resulting in unwanted computations, one possible solution could be to memoize the `paginationMeta`:

```tsx
const paginationMetaRef = React.useRef<GridPaginationMeta>();

const paginationMeta = React.useMemo(() => {
  if (
    hasNextPage !== undefined &&
    paginationMetaRef.current?.hasNextPage !== hasNextPage
  ) {
    paginationMetaRef.current = { hasNextPage };
  }
  return paginationMetaRef.current;
}, [hasNextPage]);
```

{{"demo": "ServerPaginationGridNoRowCount.js", "bg": "inline"}}

#### Estimated row count

> There could be possibilities when the accurate row count is not initially available for many reasons such as:

Author: > Would it be technically correct to say that the "Estimated row count" case is identical to the "Unknown row count", with the difference that the estimatedRowCount is used to show more detailed info in the pagination footer?

Yes!

I wanted to explain the rationale for the use case as it's kind of an edge case and not many people face this problem with the BE data fetching. But yeah, it's a bit wordy. I tried to reduce it based on your suggestions. Lemme know hyf.

> There could be possibilities when the accurate row count is not initially available for many reasons such as:

Reviewer: Nice!

There could be two possibilities after providing the initial props (as mentioned in the table above):

1. The actual row count is fetched from the server lazily and provided to the Grid using the `rowCount` prop. The `estimatedRowCount` prop will be ignored once the `rowCount` prop is set to a positive value.
2. The user has already reached the estimated last page and the actual row count is still not available, in that case, the Grid could take the help of the `hasNextPage='false'` to know that the last page is fetched and try to set a value like in the unknown row count use case.

The following example demonstrates the use of the `estimatedRowCount` and `hasNextPage` props to handle the case when the actual row count is not initially available. The actual row count is `1000` but the Grid is initially provided with an estimated row count of `100`.

The Grid keeps fetching the next page until `hasNextPage` is `false` or the actual row count is provided to the Grid lazily, you can do that by clicking the "Set Row Count" button.

> {{"demo": "ServerPaginationGridEstimated.js", "bg": "inline"}}

Reviewer: This might be out of this PR's scope, but did you consider custom pagination components in this scenario?
E.g. the one with individual page buttons in https://mui.com/x/react-data-grid/components/#pagination

In this particular example, it looks wrong: https://codesandbox.io/p/sandbox/inspiring-cherry-svvzh6

<img width="842" src="https://github.com/mui/mui-x/assets/13808724/ed65efba-8668-4681-a1c7-05fe579cb388">

- It shows only page 1 (although `estimatedRowCount` could be utilized here)
- And more importantly, the arrow buttons are disabled. This seems to be caused by https://github.com/mui/mui-x/pull/12474/files?w=1#diff-79d72ca24d1e0278032ce56430db49ef25dc59fd363d0d063912654cddc850ceR14

Reviewer: Here's the updated demo: https://codesandbox.io/p/sandbox/affectionate-violet-qfjdgw?file=%2Fsrc%2FDemo.tsx

- Index-based pagination
- Cursor-based pagination

:::info
Check out [Selection—Usage with server-side pagination](/x/react-data-grid/row-selection/#usage-with-server-side-pagination) for more details.
:::

### Index-based pagination

The index-based pagination is the most common type of pagination. It follows the same pattern as the client-side pagination (page, pageSize), but the data is fetched from the server.

To enable server-side pagination, you need to:

- Set the `paginationMode` prop to `server`
- Use the `onPaginationModelChange` prop to react to the page changes and load the data from the server

The server-side pagination can be further categorized into sub-types based on the availability of the total number of rows or `rowCount`.

The Grid uses the `rowCount` to calculate the number of pages and to show the information about the current state of the pagination in the footer.
You can provide the `rowCount` in one of the following ways:

- Initialize - use `initialState.pagination.rowCount` prop to initialize the `rowCount`
- Control - use the `rowCount` prop along with `onRowCountChange` to control the `rowCount` and reflect the changes when the row count is updated

> - Set using the `apiRef` - use the `apiRef.current.setRowCount` method to set the `rowCount` after the Grid is initialized

Author: This [way](https://www.notion.so/mui-org/Writing-style-guide-2a957a4168a54d47b14bae026d06a24b?pvs=4#e6b1ebbe81fd48de92eea2c30b0d3d1d) of listing different radio items seemed a bit better looking, I added a similar. Let me know how you feel.

```jsx
const rowCountRef = React.useRef(pageInfo?.totalRowCount || 0);

const rowCount = React.useMemo(() => {
  if (pageInfo?.totalRowCount !== undefined) {
    rowCountRef.current = pageInfo.totalRowCount;
  }
  return rowCountRef.current;
}, [pageInfo?.totalRowCount]);

<DataGrid rowCount={rowCount} />;
```

:::

#### Unknown row count

Pass the props to the Grid as explained in the table above to handle the case when the actual row count is unknown, as the following example demonstrates.

{{"demo": "ServerPaginationGridNoRowCount.js", "bg": "inline"}}

:::warning
The value of the `hasNextPage` variable might become `undefined` during loading if it's handled by some external fetching hook resulting in unwanted computations, one possible solution could be to memoize the `paginationMeta`:

```tsx
const paginationMetaRef = React.useRef<GridPaginationMeta>();

const paginationMeta = React.useMemo(() => {
  if (
    hasNextPage !== undefined &&
    paginationMetaRef.current?.hasNextPage !== hasNextPage
  ) {
    paginationMetaRef.current = { hasNextPage };
  }
  return paginationMetaRef.current;
}, [hasNextPage]);
```

:::

#### Estimated row count

> Estimated row count could be considered a hybrid approach of known and unknown row count cases, where the actual row count is not initially available, but an estimated row count is available.

Reviewer: ```suggestion
Estimated row count could be considered a hybrid approach that switches between the "Known row count" and "Unknown row count" use cases.
```

> Initially, when an `estimatedRowCount` is set and `rowCount='-1'`, the grid work identical to the "Unknown row count" use case, with the exception that the `estimatedRowCount` is used to show more detailed information in the pagination footer.

Reviewer: ```suggestion
Initially, when an `estimatedRowCount` is set and `rowCount={-1}`, the Data Grid behaves as in the "Unknown row count" use case, but with the `estimatedRowCount` value shown in the pagination footer.
```

> Once the current row count exceeds the `estimatedRowCount`, the Grid will ignore the `estimatedRowCount` and completely work as the "Unknown row count" use case.

Reviewer: ```suggestion
If the number of rows loaded exceeds the `estimatedRowCount`, the Data Grid ignores the `estimatedRowCount` and the behavior is identical to the "Unknown row count" use case.
```

> Meanwhile, if the `rowCount` is set to a valid value (either by updating the `rowCount` prop or setting `hasNextPage='false'`), the Grid will shift to the "Known row count" use case.

Reviewer: ```suggestion
When the `hasNextPage` returns `false` or `rowCount` is set to a positive number, the Data Grid switches to the "Known row count" behavior.
```

> The following example demonstrates this scenario. The actual row count is `1000` but the Grid is initially provided with an estimated row count of `100`. You can set the `rowCount` value by hitting the "Set Row Count" button.

Reviewer: ```suggestion
In the following example, the actual row count is `1000` but the Data Grid is initially provided with `estimatedRowCount={100}`.
You can set the `rowCount` to the actual row count by pressing the "Set Row Count" button.
```

{{"demo": "ServerPaginationGrid.js", "bg": "inline"}}

#### Unknown row count

When the row count is not available, you can use the following steps to enable server-side pagination:

- Set the prop `paginationMode` to `server`
- Set the `rowCount` as `-1` using one of the above mentioned ways to let the Grid know that the row count is not available
- Since the `rowCount` is unknown, the Grid needs to know whether more records are available on the server. Pass the `paginationMeta.hasNextPage` boolean property to indicate whether more records are available.

> - Use the `onPaginationModelChange` prop to react to the `paginationModel` changes

Reviewer: These sections are a bit hard to read, how about we organize everything into a table, and keep shorter descriptions for each of the scenario?
If I understand this correctly, the table should look like this:

|                              | Known row count | Unknown row count | Estimated row count |
| :--------------------------- | :-------------- | :---------------- | :------------------ |
| `rowCount`                   | `number`        | `-1`              | `-1`                |
| `paginationMeta.hasNextPage` | —               | `boolean`         | `boolean`           |
| `estimatedRowCount`          | —               | —                 | `number`            |

<details>
<summary>Raw markdown</summary>

```
|                              | Known row count | Unknown row count | Estimated row count |
| :--------------------------- | :-------------- | :---------------- | :------------------ |
| `rowCount`                   | `number`        | `-1`              | `-1`                |
| `paginationMeta.hasNextPage` | —               | `boolean`         | `boolean`           |
| `estimatedRowCount`          | —               | —                 | `number`            |
```
</details>

What do you think?

Reviewer: The `paginationMode` and `onPaginationModelChange` can be covered once in the "Index-based pagination" section too.

Author: I agree it's a bit cluttered, thanks for the suggestions. I'll try to accommodate them.

Author: @cherniavskii I did some revamping thanks to your suggestions, however, it still doesn't feel super clear readability-wise. 😞 

I am thinking of dividing this page into sub-pages. For now:

- Pagination
  - Overview
  - Server-side pagination
  - Recipes
  
The recipes section could be started with this example: https://mui.com/x/react-data-grid/pagination/#custom-pagination-ui, we could add more later.

Would it be an overkill though? Wdyt?

The `hasNextPage` must not be set to `false` until there are _actually_ no records left to fetch, because when `hasNextPage` becomes `false`, the Grid considers this as the last page and tries to set the `rowCount` value to a finite value.

Relying on this behavior of the Grid should be the last resort though, the preferred way is to set the `rowCount` to the actual number of rows fetched from the server as soon as the information is available since the Grid cannot guess if the rows in the last page are less than the page size, and the incorrect number might be displayed in the pagination UI.

> For example the if the row count is `15` and the page size is `10`, on the last page, when `hasNextPage` becomes false the Grid will set the row count to be `page (2) * pageSize (10) = (20)` showing `10-20 of 20` which is incorrect, the correct value should be `10-15 of 15`.

Reviewer: Do we really need this warning?
I think this is acceptable behavior when the row count is unknown, we don't have to discourage the users to do this if this is their use case.

One thing to clarify here – the incorrect number might appear only while fetching of the last page, and once the page is fetched – we update it with the actual number of rows (all of them have been fetched). Is this correct?

Author: > the incorrect number might appear only while fetching of the last page, and once the page is fetched – we update it with the actual number of rows (all of them have been fetched). Is this correct?

@cherniavskii Now, yes (done in https://github.com/mui/mui-x/pull/12474/commits/4f8f45dff590622d1284677a46c941c9cb230c5d), thanks for pointing it out.

Previously I was aligned to off-load this responsibility to the user (that they explicitly have to set the updated `rowCount` prop especially if the last page size on server is less that `paginationModel.pageSIze`), because loading could be set to `false` in one render cycle while rows might be updated in multiple render cycles or at a later point in time depending on how the user is fetching them, so it didn't seem safe to make it a default behavior.

However, on a rethought, it seemed fine to me (as well as a better DX) since it could be opted out by controlling the `rowCount` prop.

Removed the warning as well, since the behavior is pretty much self-explanatory to me now. Let me know if you feel like adding a couple of line may still help.

Relying on this behavior of the Grid should be the last resort though, the preferred way is to set the `rowCount` to the actual number of rows fetched from the server as soon as the information is available since the Grid cannot guess if the rows in the last page are less than the page size, and the incorrect number might be displayed in the pagination UI.

For example the if the row count is `15` and the page size is `10`, on the last page, when `hasNextPage` becomes false the Grid will set the row count to be `page (2) * pageSize (10) = (20)` showing `10-20 of 20` which is incorrect, the correct value should be `10-15 of 15`.

The value of the `hasNextPage` variable might become `undefined` during loading if it's handled by some external fetching hook resulting in flickering of the pagination UI, one possible solution could be to memoize the `paginationMeta`:

```tsx
const paginationMetaRef = React.useRef<GridPaginationMeta>();

const paginationMeta = React.useMemo(() => {
  if (
    hasNextPage !== undefined &&
    paginationMetaRef.current?.hasNextPage !== hasNextPage
  ) {
    paginationMetaRef.current = { hasNextPage };
  }
  return paginationMetaRef.current;

> }, [hasNextPage]);

Reviewer: This approach looks better than `useEffect` in the "Known row count" section 👍🏻
Would you mind updating it there?

Author: Not at all, seems a nice idea 👍
